## Exercise 4.9 -  Intro to Data Visualization with Python Task- Part I

### Step 3: Import your analysis libraries, as well as your new customer data set as a dataframe.

In [1]:
# import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# assign the folder path to the variable 'path'

path=r"C:\Users\19452\Desktop\Seema\Careerfoundry-Data Analytics Course\Work Area\Python Fundamentals\Nov-2023 Instacart Basket Analysis\02 Data"

In [3]:
# import customer dataset as a dataframe from original Data folder

df_customer=pd.read_csv(os.path.join(path,'Original Data','customers.csv'))

### Step 4: Wrangle the data 

In [4]:
# Look at the columns in df_customer dataframe

df_customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
# number of rows and columns

df_customer.shape

(206209, 10)

In [6]:
# Rename columns with illogical names
# for maintaining uniformity, change case of all columns to small

df_customer.rename(columns={'First Name':'first_name','Surnam':'surname','Gender':'gender','STATE':'state','Age':'age','n_dependants':'no_of_dependants','fam_status':'family_status'},inplace=True)


In [7]:
# check for the renamed columns

df_customer.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   no_of_dependants family_status  income  
0                 3       married  165665  
1                 0        single   59285  
2                 2       married   99568  
3                 0        single   42049  
4                 1       married   40374

In [8]:
# get statistics on columns

df_customer.describe()

user_id            age  no_of_dependants         income
count  206209.000000  206209.000000     206209.000000  206209.000000
mean   103105.000000      49.501646          1.499823   94632.852548
std     59527.555167      18.480962          1.118433   42473.786988
min         1.000000      18.000000          0.000000   25903.000000
25%     51553.000000      33.000000          0.000000   59874.000000
50%    103105.000000      49.000000          1.000000   93547.000000
75%    154657.000000      66.000000          3.000000  124244.000000
max    206209.000000      81.000000          3.000000  593901.000000

In [9]:
# Check the data types of all the columns

df_customer.dtypes

user_id              int64
first_name          object
surname             object
gender              object
state               object
age                  int64
date_joined         object
no_of_dependants     int64
family_status       object
income               int64
dtype: object

In [10]:
# convert the data type of 'user_id' from integer to string as id fields are not used for any numberic calculations

df_customer['user_id']=df_customer['user_id'].astype('str')

In [11]:
# check the data type of 'user_id' column after it is converted to string

df_customer['user_id'].dtype

dtype('O')

In [12]:
# In Python, date and time are not data types of their own, 
# but a module named DateTime in Python can be imported to work with the date as well as time. 

import datetime

In [13]:
# convert 'date_joined' column data type from object to date

df_customer['date_joined']=pd.to_datetime(df_customer['date_joined'])

In [14]:
#  Change the data format to mm-dd-yyyy

df_customer['date_joined'].dt.strftime('%m-%d-%Y')

0         01-01-2017
1         01-01-2017
2         01-01-2017
3         01-01-2017
4         01-01-2017
             ...    
206204    04-01-2020
206205    04-01-2020
206206    04-01-2020
206207    04-01-2020
206208    04-01-2020
Name: date_joined, Length: 206209, dtype: object

In [15]:
# Check the data types of all the columns

df_customer.dtypes

user_id                     object
first_name                  object
surname                     object
gender                      object
state                       object
age                          int64
date_joined         datetime64[ns]
no_of_dependants             int64
family_status               object
income                       int64
dtype: object

### Step 5 - Complete the fundamental data quality and consistency checks 

#### first_name seems to have mixed data types.  The current data type of first_name is object which is suitable for the
#### column.  So does not need changing anything.

In [16]:
# See if there are any mixed data types

for col in df_customer.columns.tolist():
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customer[weird]) > 0:
    print (col)

C:\Users\19452\AppData\Local\Temp\ipykernel_35692\3497215333.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\19452\AppData\Local\Temp\ipykernel_35692\3497215333.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\19452\AppData\Local\Temp\ipykernel_35692\3497215333.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\19452\AppData\Local\Temp\ipykernel_35692\3497215333.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis =

first_name


C:\Users\19452\AppData\Local\Temp\ipykernel_35692\3497215333.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\19452\AppData\Local\Temp\ipykernel_35692\3497215333.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\19452\AppData\Local\Temp\ipykernel_35692\3497215333.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)


#### No Duplicates found

In [17]:
# check for duplicates

df_customer_duplicates=df_customer[df_customer.duplicated()]

df_customer_duplicates

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, no_of_dependants, family_status, income]
Index: []

#### There are missing values in first_name column

In [18]:
# check for missing values

df_customer.isnull().sum()

user_id                 0
first_name          11259
surname                 0
gender                  0
state                   0
age                     0
date_joined             0
no_of_dependants        0
family_status           0
income                  0
dtype: int64

#### There are 11259 values missing in first_name column.
#### Since keeping blank first_names does not affect the analysis, no action needed

In [19]:
# Look at the rows where first_name is blank

df_customer[df_customer['first_name'].isnull()]

user_id first_name   surname  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26  2017-01-01   
73       13738        NaN     Frost  Female     Louisiana   39  2017-01-01   
82       89996        NaN    Dawson  Female        Oregon   52  2017-01-01   
99       96166        NaN   Oconnor    Male      Oklahoma   51  2017-01-01   
105      29778        NaN    Dawson  Female          Utah   63  2017-01-01   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28  2020-03-31   
206044  200799        NaN  Copeland  Female        Hawaii   52  2020-04-01   
206090  167394        NaN     Frost  Female        Hawaii   61  2020-04-01   
206162  187532        NaN     Floyd  Female    California   39  2020-04-01   
206171  116898        NaN   Delgado  Female      Colorado   23  2020-04-01   

        no_of_dependants family_status  income  
53                     2       married   41709  
73                     0        single   82518  
82                     3       married  117099  
99                     1       married  155673  
105                    3       married  151819  
...                  ...           ...     ...  
206038                 3       married   87783  
206044                 2       married  108488  
206090                 1       married   45275  
206162                 0        single   56325  
206171                 2       married   59222  

[11259 rows x 10 columns]

In [20]:
# Over all information on the updated dataframe

df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   user_id           206209 non-null  object        
 1   first_name        194950 non-null  object        
 2   surname           206209 non-null  object        
 3   gender            206209 non-null  object        
 4   state             206209 non-null  object        
 5   age               206209 non-null  int64         
 6   date_joined       206209 non-null  datetime64[ns]
 7   no_of_dependants  206209 non-null  int64         
 8   family_status     206209 non-null  object        
 9   income            206209 non-null  int64         
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 15.7+ MB


### Step 6 - Combine your customer data with the rest of your prepared Instacart data.

In [21]:
# import ords_prods_merge.pkl dataset as a dataframe in order to merge with customer dataset

df_ords_prods_merge=pd.read_pickle(os.path.join(path,'Prepared Data','ords_prods_merge.pkl'))

In [22]:
# information on the imported dataset

df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32641279 entries, 0 to 32641278
Data columns (total 23 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                float64 
 1   user_id                 float64 
 2   order_number            float64 
 3   orders_day_of_week      float64 
 4   order_hour_of_day       float64 
 5   days_since_prior_order  float64 
 6   product_id              float64 
 7   add_to_cart_order       float64 
 8   reordered               float64 
 9   product_name            object  
 10  aisle_id                float64 
 11  department_id           float64 
 12  prices                  float64 
 13  merge_flag              category
 14  price_range_loc         object  
 15  Busiest days            object  
 16  busiest_period_of_day   object  
 17  max_order               float64 
 18  loyalty_flag            object  
 19  average spent           float64 
 20  spending flag           object  
 21  regula

In [23]:
# user_id column is common in both the datasets, but the data types are different.
# to make them compatible to merge, converting user_id with object data type to float64 in df_customer dataframe

df_customer['user_id']=df_customer['user_id'].astype('float64')

In [24]:
df_customer.dtypes

user_id                    float64
first_name                  object
surname                     object
gender                      object
state                       object
age                          int64
date_joined         datetime64[ns]
no_of_dependants             int64
family_status               object
income                       int64
dtype: object

In [25]:
df_customer.head(5)

user_id first_name   surname  gender       state  age date_joined  \
0   26711.0    Deborah  Esquivel  Female    Missouri   48  2017-01-01   
1   33890.0   Patricia      Hart  Female  New Mexico   36  2017-01-01   
2   65803.0    Kenneth    Farley    Male       Idaho   35  2017-01-01   
3  125935.0   Michelle     Hicks  Female        Iowa   40  2017-01-01   
4  130797.0        Ann   Gilmore  Female    Maryland   26  2017-01-01   

   no_of_dependants family_status  income  
0                 3       married  165665  
1                 0        single   59285  
2                 2       married   99568  
3                 0        single   42049  
4                 1       married   40374

In [26]:
# check whether the user_id data type is changed

df_ords_prods_merge.dtypes

order_id                   float64
user_id                    float64
order_number               float64
orders_day_of_week         float64
order_hour_of_day          float64
days_since_prior_order     float64
product_id                 float64
add_to_cart_order          float64
reordered                  float64
product_name                object
aisle_id                   float64
department_id              float64
prices                     float64
merge_flag                category
price_range_loc             object
Busiest days                object
busiest_period_of_day       object
max_order                  float64
loyalty_flag                object
average spent              float64
spending flag               object
regularity                 float64
order frequency flag        object
dtype: object

In [27]:
df_ords_prods_merge.head(5)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329.0      1.0           1.0                 2.0                8.0   
1  2398795.0      1.0           2.0                 3.0                7.0   
2   473747.0      1.0           3.0                 3.0               12.0   
3  2254736.0      1.0           4.0                 4.0                7.0   
4   431534.0      1.0           5.0                 4.0               15.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   

  product_name  ...  merge_flag    price_range_loc    Busiest days  \
0         Soda  ...        both  Mid-range product     Regular day   
1         Soda  ...        both  Mid-range product  Least busy day   
2         Soda  ...        both  Mid-range product  Least busy day   
3         Soda  ...        both  Mid-range product  Least busy day   
4         Soda  ...        both  Mid-range product  Least busy day   

  busiest_period_of_day max_order      loyalty_flag average spent  \
0           Most orders      11.0  Regular customer      6.367797   
1         Fewest orders      11.0  Regular customer      6.367797   
2           Most orders      11.0  Regular customer      6.367797   
3         Fewest orders      11.0  Regular customer      6.367797   
4           Most orders      11.0  Regular customer      6.367797   

   spending flag regularity  order frequency flag  
0    Low spender       20.0      Regular customer  
1    Low spender       20.0      Regular customer  
2    Low spender       20.0      Regular customer  
3    Low spender       20.0      Regular customer  
4    Low spender       20.0      Regular customer  

[5 rows x 23 columns]

In [28]:
# Merge df_customer dataframe with df_ords_prods_merge dataframe using user_id as the common key

df_allmerge=df_ords_prods_merge.merge(df_customer,on='user_id',indicator=True)

In [29]:
# look at the combined dataset

df_allmerge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329.0      1.0           1.0                 2.0                8.0   
1   2398795.0      1.0           2.0                 3.0                7.0   
2    473747.0      1.0           3.0                 3.0               12.0   
3   2254736.0      1.0           4.0                 4.0                7.0   
4    431534.0      1.0           5.0                 4.0               15.0   
5   3367565.0      1.0           6.0                 2.0                7.0   
6    550135.0      1.0           7.0                 1.0                9.0   
7   3108588.0      1.0           8.0                 1.0               14.0   
8   2295261.0      1.0           9.0                 1.0               16.0   
9   2550362.0      1.0          10.0                 4.0                8.0   
10  2539329.0      1.0           1.0                 2.0                8.0   
11  2539329.0      1.0           1.0                 2.0                8.0   
12  2398795.0      1.0           2.0                 3.0                7.0   
13   473747.0      1.0           3.0                 3.0               12.0   
14  2254736.0      1.0           4.0                 4.0                7.0   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN       196.0                1.0        0.0   
1                     15.0       196.0                1.0        1.0   
2                     21.0       196.0                1.0        1.0   
3                     29.0       196.0                1.0        1.0   
4                     28.0       196.0                1.0        1.0   
5                     19.0       196.0                1.0        1.0   
6                     20.0       196.0                1.0        1.0   
7                     14.0       196.0                2.0        1.0   
8                      0.0       196.0                4.0        1.0   
9                     30.0       196.0                1.0        1.0   
10                     NaN     14084.0                2.0        0.0   
11                     NaN     12427.0                3.0        0.0   
12                    15.0     12427.0                3.0        1.0   
13                    21.0     12427.0                2.0        1.0   
14                    29.0     12427.0                2.0        1.0   

                               product_name  ...  first_name  surname  gender  \
0                                      Soda  ...       Linda   Nguyen  Female   
1                                      Soda  ...       Linda   Nguyen  Female   
2                                      Soda  ...       Linda   Nguyen  Female   
3                                      Soda  ...       Linda   Nguyen  Female   
4                                      Soda  ...       Linda   Nguyen  Female   
5                                      Soda  ...       Linda   Nguyen  Female   
6                                      Soda  ...       Linda   Nguyen  Female   
7                                      Soda  ...       Linda   Nguyen  Female   
8                                      Soda  ...       Linda   Nguyen  Female   
9                                      Soda  ...       Linda   Nguyen  Female   
10  Organic Unsweetened Vanilla Almond Milk  ...       Linda   Nguyen  Female   
11                      Original Beef Jerky  ...       Linda   Nguyen  Female   
12                      Original Beef Jerky  ...       Linda   Nguyen  Female   
13                      Original Beef Jerky  ...       Linda   Nguyen  Female   
14                      Original Beef Jerky  ...       Linda   Nguyen  Female   

      state age date_joined no_of_dependants  family_status income  _merge  
0   Alabama  31  2019-02-17                3        married  40423    both  
1   Alabama  31  2019-02-17                3        married  40423    both  
2   Alabama  31  2019-02-17                3        married

In [30]:
df_allmerge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32641268 entries, 0 to 32641267
Data columns (total 33 columns):
 #   Column                  Dtype         
---  ------                  -----         
 0   order_id                float64       
 1   user_id                 float64       
 2   order_number            float64       
 3   orders_day_of_week      float64       
 4   order_hour_of_day       float64       
 5   days_since_prior_order  float64       
 6   product_id              float64       
 7   add_to_cart_order       float64       
 8   reordered               float64       
 9   product_name            object        
 10  aisle_id                float64       
 11  department_id           float64       
 12  prices                  float64       
 13  merge_flag              category      
 14  price_range_loc         object        
 15  Busiest days            object        
 16  busiest_period_of_day   object        
 17  max_order               float64       
 18  

### Step 8 - Export the new dataframe as pickle file

In [31]:
# Export df_allmerge dataframe to the prepare data folder as 'allmergedata_1.pkl'

df_allmerge.to_pickle(os.path.join(path,'Prepared Data','allmergedata_1.pkl'))